### ISSUE TITLE
Grouping services to one endpoint

### ISSUE THREAD

**Issue by vlesierse** [magneticio/vamp#126]
As part of my application I would like to expose two services on the same port and route them based on their paths using a `path_beg` ACL.

- http://application.com/api/a
- http://application.com/api/b

Of course I could put an API Proxy in front of the Vamp router, but it would be nicer if Vamp can support this on it's own to reduce network overhead.

```
                                    /--> Service A
Client --> API Proxy --> Vamp Router
                                    \--> Service B
```

If I'm correct the Vamp router it self can support a scenario like this already, but Vamp gives me a hard time to actually define this using a blueprint.

## Option 1
Use one cluster with multiple breeds

```  yaml
name: application
endpoints:
    service.port: 80
clusters:
    application-api:
        services:
            -   breed:
                    name: service-a
                    ports:
                        port: 8080/http
                routing:
                    filters:
                        - condition: path_beg /api/a
            -   breed:
                    name: service-b
                    ports:
                        port: 8888/http
                routing:
                    filters:
                        - condition: path_beg /api/b                
```
This options currently works without making changes to Vamp. However the weights and filters apply to all the breeds within the service which make canary releasing difficult to impossible.

How should it be possible to specify that `service-a:1.0` and `service-a:2.0` gets 50% but also `service-b:1.0` and `service-b:2.0`  

## Option 2
Grouping endpoints based on ports

```  yaml
name: application
endpoints:
    service-a.port: 80
    service-b.port: 80
clusters:
    service-a:
        services:
            breed:
                name: service-a
                ports:
                    port: 8080/http
            routing:
                filters:
                    - condition: path_beg /api/a
    service-b:
        services:
            breed:
                name: service-b
                ports:
                    port: 8080/http
            routing:
                filters:
                    - condition: path_beg /api/b
                
```
This options currently doesn't work because it will create a new port for each entry in endpoints even when the same port is specified.

Changing vamp by grouping the endpoints together using the name and port would solve this although it will still be hard to express the filter conditions to redirect the requests to a specific service.

### Option 3
Express filtering and grouping on endpoints

```  yaml
name: application
endpoints:
    application-api:
        port: 80/http
        services:
            service-a.port:
                - condition: path_beg /api/a
            service-b.port:
                - condition: path_beg /api/b
clusters:
    service-a:
        services:
            breed:
                name: service-a
                ports:
                    port: 8080/http
    service-b:
        services:
            breed:
                name: service-b
                ports:
                    port: 8888/http
```
Making it possible to describe the endpoints and specify the routing to the services underneath could enable Vamp to support this scenario.

**Comment by tnolet** [comment]
Hi @vlesierse . Regardless of the implementation and DSL syntax, what you effectively want is pre-service routing/proxying: routing that is completely independent from any filters/weights used for canary releasing. This is much like an API-gateway would function. You are correct that Vamp Router would totally allow this, we just have to enable it in the DSL.

We will make this into a feature proposal as I also noticed this pattern in the project discussed here: https://github.com/magneticio/vamp/issues/122 They had a nice setup where the actual services are dependencies of a top level nginx instance.

As a workaround for now, you could follow a similar pattern:
1) Have a customer facing nginx as an endpoint that proxy passes to underlying services. See  the accompanying nginx config file.https://github.com/zutherb/AppStash/blob/master/microservice/frontend/catalog/docker/default.tmpl
2)Then declare the services as dependencies and use the environment variables to hook them up to nginx

also check their blueprint at https://github.com/zutherb/AppStash/blob/master/vamp/shop.yml 
it gives a nice suggestion how this will work.

**Comment by tnolet** [comment]
@dragoslav What's your opinion on the idea of option 3?

**Comment by dragoslav** [comment]
The option 3 is aligned with changes we already planned/proposed. At the moment DSL endpoint are the simple proxies and adding filter & weight support is something we can easily implement (using basically already existing implementation). We can plan this feature for the coming 2-3 weeks.

**Comment by tnolet** [comment]
@vlesierse does the explanation make sense?

**Comment by vlesierse** [comment]
@tnolet @dragoslav Yeah, it totally makes sense. Thnx. For the moment we have a workaround and I would like to be happy to help/test the support from the DSL.

**Comment by dragoslav** [comment]
Closed as it is a part of #167
